modules/bullet/space_bullet.cpp

```diff
@@ -152,6 +152,16 @@ int BulletPhysicsDirectSpaceState::intersect_shape(const RID &p_shape, const Tra
 }
 
 bool BulletPhysicsDirectSpaceState::cast_motion(const RID &p_shape, const Transform &p_xform, const Vector3 &p_motion, float p_margin, float &r_closest_safe, float &r_closest_unsafe, const Set<RID> &p_exclude, uint32_t p_collision_mask, bool p_collide_with_bodies, bool p_collide_with_areas, ShapeRestInfo *r_info) {
+
+	r_closest_safe = 1.f;
+	r_closest_unsafe = 1.f;
+
+	btVector3 bt_motion;
+	G_TO_B(p_motion, bt_motion);
+
+	if (bt_motion.fuzzyZero())
+		return true;
```

Comment:
Returns true imply that the query got executed, but this is not the case.

If we want to return `true`, we have to forward this call to the `shape_cast` query and so fill `r_info` correctly.

I prefer to leave this API pure, returning `false`; so the user can decide freely what to do next.
In case we return `false` it's necessary don't alter the content of:

```cpp
r_closest_safe = 1.f;
r_closest_unsafe = 1.f;
```

Reply:
My understanding of the purpose of the function is to test whether a motion is possible without a collision. If it is possible to move without a collision, it returns true and sets the values of `closest_safe` and `closest_unsafe` to `1.f`. If it is not possible to move without a collision, it returns false and sets the value of `closest_safe` to the proportion of the move that is possible without a collision and `closest_unsafe` to the proportion of the move that would result in a collision.

I think most users would expect that, if the motion length is zero, the "motion" is always possible and the function should return true. The only time I would expect a zero length "motion" not to be possible without a collision, is if the object is already overlapping with another object. First, this is unlikely. Second, what would a user be expected to do in this case?

Regardless of whether a zero length motion should return true or false, the function is expected to initialise the `closest_safe` and `closest_unsafe` values; as seen by the lack of assigning default values, fixed with the other changes in this commit.

Therefore, I think my approach is the correct one: there is no need to check if a zero length motion would cause a collision, and the function can simply return true and set the values of `closest_safe` and `closest_unsafe` to `1.f`.

Reply:
>  This function returns `true` or `false` depending if was or wasn't possible execute the query.
Then, the actual motion is controlled by: `r_closest_safe` and `r_closest_unsafe`.

The current behaviour (should the compiler and hardware support IEEE 754 and not crash when dividing a float by zero) is for a zero length `p_motion` vector to return `true` and for the values of `r_closest_safe` and `r_closest_unsafe` to be set to the default `1.0`.

In `btCollisionWorld::convexSweepTest()`, when creating the `btSingleSweepCallback`, the division of the zero vector by its zero length when calling `normalized()` results in the `rayDir` vector being all `NaN`. These `NaN`s are carried around and copied until the following check of a `Nan` < `Nan` fails:
https://github.com/godotengine/godot/blob/817165b96cd04adb29e4eef797f9dcc6dd9d022f/thirdparty/bullet/LinearMath/btAabbUtil2.h#L114
This failure during the `btDbvtBroadphase::rayTest()` suggests (correctly or incorrectly) that nothing is nearby and there is no need to perform further tests.

The effect is for `btResult` not to be updated, and for `btResult.hasHit()` to return `false`. The `r_closest_safe` and `r_closest_unsafe` values are not changed from their default `1.0`, and `cast_motion()` returns `true`.

Therefore, my approach doesn't change the current behaviour (and I still think it is the expected behaviour).


>or the intermediate one:
https://github.com/godotengine/godot/blob/74977277fac258060b111b31ec57ed2fa8b63398/modules/bullet/space_bullet.cpp#L157-L162

Aside from the point that only supporting convex shapes should probably be addressed as opposed to simply printing an error and returning false, it should also probably be setting values for `closest_safe` and `closest_unsafe`.

[`PhysicsDirectSpaceState::cast_motion()`](https://docs.godotengine.org/en/3.1/classes/class_physicsdirectspacestate.html#class-physicsdirectspacestate-method-cast-motion) currently uses a return value of false to set them to 0:
https://github.com/godotengine/godot/blob/817165b96cd04adb29e4eef797f9dcc6dd9d022f/servers/physics_server.cpp#L313-L320

Therefore I suggest we do the same.

Reply:
Added `r_closest_safe = 0.f;` and `r_closest_unsafe = 0.f;` before returning `false` in `if(!btShape->isConvex())`.

Reply:
Let me reiterate my sentence; even with motion equal to __0__ I expect that the function is executed with 0 motion and so it fills the `r_info` properly with colliding information (if any).

My proposals to fix this it are:
- Returns `false` so the user know that it's not possible to process this query with 0 movement and he will address it in another way.
- Implicitly call the `intersect_shape` function and correctly fill `r_motion`

> Aside from the point that only supporting convex shapes should probably be addressed as opposed to simply printing an error and returning false

Move a convex shape is not supported in Bullet, for this reason it simply return `false` and the error message.

Reply:
To make the discussion easier I've tabulated the options:
Option 1:
<table>
	<tr>
		<td>returns</td>
		<td>closest_safe</td>
		<td>closest_unsafe</td>
	</tr>
	<tr>
		<td>TRUE</td>
		<td>1</td>
		<td>1</td>
	</tr>
</table>

Option 2:
<table>
	<tr>
		<td>returns</td>
		<td>closest_safe</td>
		<td>closest_unsafe</td>
	</tr>
	<tr>
		<td>FALSE</td>
		<td>unset</td>
		<td>unset</td>
	</tr>
</table>

If we compare these options with the current behaviour in the two Physics options, we see a strong correlation:
Current Bullet Physics:
<table>
	<tr>
		<td>returns</td>
		<td>closest_safe</td>
		<td>closest_unsafe</td>
	</tr>
	<tr>
		<td>TRUE</td>
		<td>1</td>
		<td>1</td>
	</tr>
</table>

Current Godot Physics:
<table>
	<tr>
		<td>returns</td>
		<td>closest_safe</td>
		<td>closest_unsafe</td>
	</tr>
	<tr>
		<td>FALSE</td>
		<td>unset</td>
		<td>unset</td>
	</tr>
</table>

However, this is not the full picture. The above results are for motion = 0, and there is an issue with Bullet Physics with motion = 0 that this PR is trying to fix. What we need is for motion = 0 to be the same as the limit motion->0, or alternatively the limit possible safe motion->0 for any motion > 0. The following are the results for motion > 0 with possible safe motion = 0 created by placing the object to be cast inside another object.

Current Bullet Physics with motion > 0:
<table>
	<tr>
		<td>Possible safe</td>
		<td>returns</td>
		<td>closest_safe</td>
		<td>closest_unsafe</td>
	</tr>
	<tr>
		<td>1</td>
		<td>TRUE</td>
		<td>1</td>
		<td>1</td>
	</tr>
	<tr>
		<td>0</td>
		<td>TRUE</td>
		<td>0</td>
		<td>0</td>
	</tr>
</table>

Current Godot Physics with motion > 0:
<table>
	<tr>
		<td>Possible safe</td>
		<td>returns</td>
		<td>closest_safe</td>
		<td>closest_unsafe</td>
	</tr>
	<tr>
		<td>1</td>
		<td>TRUE</td>
		<td>1</td>
		<td>1</td>
	</tr>
	<tr>
		<td>0</td>
		<td>FALSE</td>
		<td>unset</td>
		<td>unset</td>
	</tr>
</table>

These results reveal a problem with the Godot Physics version when motion > 0 and the possible safe motion is 0; so the current Godot Physics results are not useful either.

Therefore, my revised suggestion is to use the limit as possible motion->0 for motion > 0 provided by Bullet Physics for setting the closest_safe and closest_unsafe values i.e. setting both to 0. Then, since returning true or false will have the same outcome, I'm happy for the function to return false. To summarise, when motion = 0:
<table>
	<tr>
		<td>returns</td>
		<td>closest_safe</td>
		<td>closest_unsafe</td>
	</tr>
	<tr>
		<td>FALSE</td>
		<td>0</td>
		<td>0</td>
	</tr>
</table>

Reply:
More I think to this function more I'm convinced that we should implicitly call `intersect_shape` because I would expect that the collision check is done even with `motion == 0`.

Would be good to hear the voice of some users to understand if we should just return `false` or forward the call to `intersect_shape`.

@akien-mga any idea?

Reply:
What's the state of this?

Reply:
I think this PR is awaiting a resolution on this change request, which in turn is awaiting an agreement on what the `cast_motion()` should return when `p_motion = Vector3()`.

Currently the PR sets `r_closest_safe = 0.f` and `r_closest_unsafe = 0.f` and returns false.